tango/step_caches/remote_step_cache.py

```diff
@@ -0,0 +1,159 @@
+import tempfile
+from pathlib import Path
+from typing import Any, Union
+
+from tango.common.aliases import PathOrStr
+from tango.common.exceptions import TangoError
+from tango.common.file_lock import FileLock
+from tango.common.params import Params
+from tango.common.remote_utils import RemoteConstants
+from tango.step import Step
+from tango.step_cache import CacheMetadata
+from tango.step_caches.local_step_cache import LocalStepCache
+from tango.step_info import StepInfo
+
+logger = logging.getLogger(__name__)
+
+
+class RemoteNotFoundError(TangoError):
+    """
+    Classes inheriting from the RemoteStepCache should raise this if a step result object is not found.
+    """
+
+
+# This class inherits from `LocalStepCache` to benefit from its in-memory "weak cache" and "strong cache",
+# but it handles saving artifacts to disk a little differently.
+class RemoteStepCache(LocalStepCache):
+    """
+    This is a :class:`~tango.step_cache.StepCache` that's used by :class:`RemoteWorkspace`.
+    It stores the results of steps on some RemoteWorkspace.
+
+    It also keeps a limited in-memory cache as well as a local backup on disk, so fetching a
+    step's resulting subsequent times should be fast.
+
+    .. tip::
+        All remote step caches inherit from this.
+    """
+
+    def __init__(self, local_dir: Path):
+        super().__init__(local_dir)
+
+    def _acquire_step_lock_file(self, step: Union[Step, StepInfo], read_only_ok: bool = False):
+        return FileLock(
+            self.step_dir(step).with_suffix(".lock"), read_only_ok=read_only_ok
+        ).acquire_with_updates(desc=f"acquiring step cache lock for '{step.unique_id}'")
+
+    def __contains__(self, step: Any) -> bool:
+        if isinstance(step, (Step, StepInfo)):
+            cacheable = step.cache_results if isinstance(step, Step) else step.cacheable
+            if not cacheable:
+                return False
+
+            # TODO: beaker seems to not check locally. Should it?
+            # Updated TODO: does Pete's change does the same thing?
+            key = step.unique_id
+
+            # First check if we have a copy in memory.
+            if key in self.strong_cache:
+                return True
+            if key in self.weak_cache:
+                return True
+
+            # Then check if we have a copy on disk in our cache directory.
+            with self._acquire_step_lock_file(step, read_only_ok=True):
+                if self.step_dir(step).is_dir():
+                    return True
+
+            # If not, check the remote location.
+            return self._step_result_remote(step) is not None
+        else:
+            return False
+
+    def _step_results_dir(self) -> str:
+        return RemoteConstants.STEP_RESULT_DIR
+
+    def _step_result_remote(self, step: Union[Step, StepInfo]):
+        raise NotImplementedError()
+
+    def _fetch_step_remote(self, step_result, target_dir: PathOrStr):
+        raise NotImplementedError()
```

Comment:
I think it would be good to annotate the return type of these methods, and also decorate them with `@abstractmethod` (from the `abc` module)